eng/native/configurecompiler.cmake

```diff
@@ -429,8 +429,10 @@ if (CLR_CMAKE_HOST_UNIX)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-class-memaccess>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-misleading-indentation>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-stringop-overflow>)
+    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-restrict>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-stringop-truncation>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-placement-new>)
+    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-mismatched-new-delete>)
```

Comment:
an example of error without this suppression:
```
  [ 54%] Building CXX object vm/CMakeFiles/cee_dac.dir/genericdict.cpp.o
  In file included from /runtime/src/coreclr/inc/clrhost.h:18,
                   from /runtime/src/coreclr/inc/check.inl:8,
                   from /runtime/src/coreclr/inc/check.h:724,
                   from /runtime/src/coreclr/inc/contract.h:237,
                   from /runtime/src/coreclr/inc/stdmacros.h:21,
                   from /runtime/src/coreclr/vm/common.h:87,
                   from /runtime/artifacts/obj/coreclr/Linux.x64.Release/vm/wks/CMakeFiles/cee_wks_core.dir/cmake_pch.hxx:5,
                   from <command-line>:
  In function ‘void DeleteArray(TYPE*) [with TYPE = ThreadLocalModule]’,
      inlined from ‘void FunctionBase<TYPE, ACQUIREF, RELEASEF>::DoRelease() [with TYPE = ThreadLocalModule*; void (* ACQUIREF)(TYPE) = DoNothing<ThreadLocalModule*>; void (* RELEASEF)(TYPE) = DeleteArray<ThreadLocalModule>]’ at /runtime/src/coreclr/inc/holder.h:716:17,
      inlined from ‘void BaseHolder<TYPE, BASE, DEFAULTVALUE, IS_NULL>::Release() [with TYPE = ThreadLocalModule*; BASE = FunctionBase<ThreadLocalModule*, DoNothing<ThreadLocalModule*>, DeleteArray<ThreadLocalModule> >; long unsigned int DEFAULTVALUE = 0; BOOL (* IS_NULL)(TYPE, TYPE) = CompareDefault<ThreadLocalModule*>]’ at /runtime/src/coreclr/inc/holder.h:293:28,
      inlined from ‘BaseHolder<TYPE, BASE, DEFAULTVALUE, IS_NULL>::~BaseHolder() [with TYPE = ThreadLocalModule*; BASE = FunctionBase<ThreadLocalModule*, DoNothing<ThreadLocalModule*>, DeleteArray<ThreadLocalModule> >; long unsigned int DEFAULTVALUE = 0; BOOL (* IS_NULL)(TYPE, TYPE) = CompareDefault<ThreadLocalModule*>]’ at /runtime/src/coreclr/inc/holder.h:261:16,
      inlined from ‘BaseWrapper<ThreadLocalModule*, FunctionBase<ThreadLocalModule*, DoNothing<ThreadLocalModule*>, DeleteArray<ThreadLocalModule> >, 0, CompareDefault<ThreadLocalModule*> >::~BaseWrapper()’ at /runtime/src/coreclr/inc/holder.h:469:7,
      inlined from ‘Wrapper<ThreadLocalModule*, DoNothing<ThreadLocalModule*>, DeleteArray<ThreadLocalModule>, 0, CompareDefault<ThreadLocalModule*>, true>::~Wrapper()’ at /runtime/src/coreclr/inc/holder.h:781:7,
      inlined from ‘SpecializedWrapper<_TYPE, _RELEASEF>::~SpecializedWrapper() [with _TYPE = ThreadLocalModule; void (* _RELEASEF)(_TYPE*) = DeleteArray<ThreadLocalModule>]’ at /runtime/src/coreclr/inc/holder.h:846:5,
      inlined from ‘static ThreadLocalModule* ThreadStatics::AllocateAndInitTLM(ModuleIndex, PTR_ThreadLocalBlock, Module*)’ at /runtime/src/coreclr/vm/threadstatics.cpp:657:1:
  /runtime/src/coreclr/inc/holder.h:1029:5: error: ‘void operator delete [](void*)’ called on pointer returned from a mismatched allocation function [-Werror=mismatched-new-delete]
   1029 |     delete [] value;
        |     ^~~~~~~~~~~~~~~
  In file included from /runtime/src/coreclr/vm/methodtable.inl:18,
                   from /runtime/src/coreclr/vm/common.h:416:
  In static member function ‘static void* ThreadLocalModule::operator new(size_t, ParentModule)’,
      inlined from ‘static ThreadLocalModule* ThreadStatics::AllocateTLM(Module*)’ at /runtime/src/coreclr/vm/threadstatics.cpp:707:65,
      inlined from ‘static ThreadLocalModule* ThreadStatics::AllocateAndInitTLM(ModuleIndex, PTR_ThreadLocalBlock, Module*)’ at /runtime/src/coreclr/vm/threadstatics.cpp:649:79:
  /runtime/src/coreclr/vm/threadstatics.h:464:30: note: returned from ‘void* operator new(size_t)’
    464 |         return ::operator new(size);
        |                ~~~~~~~~~~~~~~^~~~~~
  cc1plus: all warnings being treated as errors
  make[2]: *** [vm/wks/CMakeFiles/cee_wks_core.dir/build.make:1238: vm/wks/CMakeFiles/cee_wks_core.dir/__/threadstatics.cpp.o] Error 1
  make[2]: *** Waiting for unfinished jobs....
```

Reply:
As discussed in previous update (gcc 11) about `Wno-placement-new`, while compiler analysis is getting better at detecting these, some of them are ignoable. It would be nice to comment out these two lines and fix the places it is highlighting.

Reply:
@am11 Let me see about doing that today, this seems like just being good C++ users. Are you aware of any suppressions we put in place for MSVC? I'd like to try and enable detection via MSVC if possible as well.

Reply:
Adding @mikem8361 for visibility since he has been looking at other allocator confusion in dac scenarios.

Reply:
@am11 This should be fixable in the following location by changing `NewArrayHolder` to `NewHolder`.

https://github.com/dotnet/runtime/blob/130ef3cae8b88593cbd1627d31c2c4bbf6336c40/src/coreclr/vm/threadstatics.cpp#L649

Reply:
@AaronRobinsonMSFT, thanks. I think it would be best to fix it separately since there are more places which compiler complains about.

To repro, we can delete line 434 and 435 and run:

```sh
docker run -v$(pwd):/runtime \
   mcr.microsoft.com/dotnet-buildtools/prereqs:debian-11-gcc12-amd64-20220511124845-b7a6185 \
   /runtime/build.sh clr+mono+libs.native+host.native -c Release -gcc
```
(and once with `-c Debug`) to find out all instances.

Reply:
> I think it would be best to fix it separately since there are more places which compiler complains about.

Agree.

PR to fix the specific one above: https://github.com/dotnet/runtime/pull/69762.